enhancements/installer/agent-based-installation-in-hive.md

```diff
@@ -0,0 +1,539 @@
+cluster until an actor has approved it. From a security standpoint, it is not
+OK for anyone who can access a copy of a discovery ISO and/or access the API
+where agents report themselves to implicitly have the capability to join a
+cluster as a Node.
+
+The Agent CRD will have a field in which to designate that it is approved. If
+the host was booted via the baremetal-operator, approval will be granted
+automatically. The same automation that caused the host to boot would have the
+ability to recognize the resulting Agent and mark it as approved.
+
+#### Create Cluster
+
+This scenario takes place on a hub cluster where hive and possibly RHACM are
+present. This scenario does not include centralized machine management.
+
+This scenario is an end-to-end flow that enables the user to specify everything
+up-front and then have an automated process create a cluster. Because the
+user's primary frame of reference is hardware-oriented, this flow enables them
+to specify Node attributes such as "role" with their BareMetalHost definition.
+Specifying those attributes together is particularly useful in a gitops
+approach.
+
+1. Infra Owner creates an InstallEnv resource. It can include fields such as egress proxy, NTP server, SSH public key, ... In particular it includes an Agent label selector, and a separate field of labels that will be automatically applied to Agents.
+1. Infra Owner creates BareMetalHost resources with corresponding BMC credentials. They must be labeled so that they match the selctor on the InstallEnv and must be in the same namespace as the InstallEnv.
+1. A new controller, the Baremetal Agent Controller, sees the matching BareMetalHosts and boots them using the discovery ISO URL found in the InstallEnv's status.
+1. The Agent starts up on each host and reports back to the assisted service, which creates an Agent resource in the cluster. The Agent is automatically labeled with the labels that were specified in the InstallEnv's Spec.
+1. The Baremetal Agent Controller sets the Agent's Role field in its spec to "master" or "worker" if a corresponding label was present on its BareMetalHost.
+1. The Agent is automatically marked as Approved via a field in its Spec based on being recognized as running on the known BareMetalHost.
+1. The Agent runs through the validation and inspection phases. The results are shown on the Agent's Status, and eventually a condition marks the Agent as "ready".
+1. Cluster Creator creates a ClusterDeployment describing a new cluster. It describes how many control-plane and worker agents to expect. It also includes a label selector to match Agent resources that should be part of the cluster.
+1. Cluster Creator applies a label to Agents if necessary so that they match the ClusterDeployment's selector.
+1. Once there are enough ready Agents of each role to fulfill the expected number as expressed on the ClusterDeployment, installation begins.
+
```

Comment:
Hive itself can already adopt a cluster. I think our work is to ensure that the "Day 2 Add Node Virtualmedia Multicluster" workflow will work on a cluster that was adopted. I can't think of any changes that would require in the API or workflows to support that. Maybe for the purposes of this proposal we can state that the workflow is only supported for ClusterDeployments where the platform is bare metal. I'll add that below, and let me know what you think of if there's more to this topic that you have in mind.

Reply:
I guess what I wonder is:

1. To adopt a cluster, the user will create a ClusterDeployment with `Spec.Installed = true` - what in particular will we require be set in the agentBaremetal platform and the agent install strategy?
2. I assume the common case is for users to do this with `hiveutil create-cluster --adopt` for this - so we will add agent-based provisioning support there?
3. To adopt a cluster and add workers, a user would also need to create an InstallEnv resource. Is there any difference in what's required in InstallEnv in this case? Stuff that is only needed for install would be optional?
4. What would be the constraints on the clusters we can adopt for agent-based provisioning be? Can we adopt any random cluster and use agent-based provisioning to add workers?

Reply:
(1): hiveutil is just a dev tool, so likely not the means by which anyone would adopt a cluster, presumably that would be handled in ACM somewhere. 

(2): We would need at the very least spec.provisioning.installStrategy.agentSelector and already I see a problem, that needs to move out of provisioning if it's being used for day 2, adopted clusters, which we did not provision. That probably means it should move to the Agent platform. We should discuss what else would be needed.

Reply:
The only thing we require for adding a worker on day 2 is a pointer to the API server (the agent validates that it can contact it, and then it's used to pull ignition).

Reply:
I was thinking it just moves to the hive CD platform instead of install strategy, no longer specific to the installation, but relevant for any agent based cluster.